### PR TITLE
Make pointer offset methods/intrinsics const

### PR DESCRIPTION
**Pull request by josephlr**
Implements #71499 using [the implementations from miri](https://github.com/rust-lang/miri/blob/52f5d202bdcfe8986f0615845f8d1647ab8a2c6a/src/shims/intrinsics.rs#L96-L112).

I added some tests what's allowed and what's UB. Let me know if any other cases should be added. 

CC: @RalfJung @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
Please also move https://github.com/rust-lang/miri/blob/147ea8f400de3ca529abcb5eb7b65f84a4896ae9/src/operator.rs#L98 to the miri engine, so that you don't have to repeat its logic. Miri can then call the engine method in the future.

**Comment by josephlr** [comment]
@oli-obk after this PR lands I was just going to remove all of the Miri `pointer_offset_inbounds` code. Will that not be possible?

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:283]: Better use `checked_mul`, which has no risk of accidentally using the result when there was an overflow.

File: src/test/ui/consts/offset.rs
Comment (RalfJung) [inline src/test/ui/consts/offset.rs:20]: Why not use that feature? I think tests should be as readable as possible.

But how would you use it anyway? The final value would still be compared in run-time-code, I presume?

Also, please explain how this type is used. I had to reverse engineer that from the code.

File: src/test/ui/consts/offset_ub.stderr
Comment (RalfJung) [inline src/test/ui/consts/offset_ub.stderr:24]: The error should start with a lower-case letter

**Comment by RalfJung** [comment]
> I made the implementation a single function for both offset and arith_offset.

I think this greatly decreased readability. Checked and unchecked arithmetic are very different operations, I am not sure if it is worth sharing the bit of code they have in common.

But we'll see again after you moved `pointer_offset_inbounds` to `librustc_mir/interpret/operator.rs`.

**Comment by JohnCSimon** [comment]
Ping from triage
@josephlr - can you please post your status on this PR? Thanks!

File: src/test/ui/consts/offset.rs
Comment (josephlr) [reply src/test/ui/consts/offset.rs:20]: I was trying to get this to work with `#[feature(const_compare_raw_pointers)]` but there seem to be some bugs with that feature. I keep getting a `"pointer arithmetic or comparison" needs an rfc` error (even when the feature is enabled).

My use case is just putting the pointers in static data, so I might just try to make the test cases better reflect my use-case.

**Comment by josephlr** [comment]
> Ping from triage
> @josephlr - can you please post your status on this PR? Thanks!

Status: Tried getting `#[feature(const_compare_raw_pointers)]` to work with this, but that was a dead-end. Working on cleaning up the tests and moving `pointer_offset_inbounds`.

File: src/test/ui/consts/offset.rs
Comment (josephlr) [reply src/test/ui/consts/offset.rs]: I got this working with `offset_of`, so now the test constants are just bools, which are then checked in the `main` function. I would have preferred using `assert_eq`, but that isn't supported in constants yet.

File: src/test/ui/consts/offset_ub.stderr
Comment (josephlr) [reply src/test/ui/consts/offset_ub.stderr]: Fixed.

File: src/librustc_mir/interpret/intrinsics.rs
Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs]: Fixed by separating the implementations for `offset` and `arith_offset`.

**Comment by josephlr** [comment]
@RalfJung @oli-obk this is ready again for review, sorry for the delay.

> Please also move https://github.com/rust-lang/miri/blob/147ea8f400de3ca529abcb5eb7b65f84a4896ae9/src/operator.rs#L98 to the miri engine, so that you don't have to repeat its logic. Miri can then call the engine method in the future.

I moved miri's `EvalContextExt::pointer_offset_inbounds` method to rustc's `InterpCx::ptr_offset_inbounds`. The name is slightly different so that miri won't break when this PR hits nightly.

BTW: https://github.com/rust-lang/miri/pull/1412 contains the miri cleanup

> > I made the implementation a single function for both offset and arith_offset.
> 
> I think this greatly decreased readability. Checked and unchecked arithmetic are very different operations, I am not sure if it is worth sharing the bit of code they have in common.

Looking at this again, you're correct, keeping the separate adds ~3 lines and makes the code much more readable.

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:438]:
```suggestion
        // The offset being in bounds cannot rely on "wrapping around" the address space.
        // So, first rule out overflows.
```

File: src/test/ui/consts/offset_ub.rs
Comment (RalfJung) [inline src/test/ui/consts/offset_ub.rs:16]: Please add `ERROR` annotations here to make sure this test cannot pass when these constants do not error.

**Review by RalfJung — APPROVED** [review]
Aside from the nit I raised, Miri implementation and test suite LGTM. For the glue and libcore changes I'll defer to @oli-obk.

File: src/test/ui/consts/offset_ub.rs
Comment (oli-obk) [reply src/test/ui/consts/offset_ub.rs:16]: Note that if the test passes, this will fail the test suite. You need a `// check-pass` comment or sth to make it not fail the test suite. Still, annotations are better so you see something directly in the test

**Review by oli-obk — APPROVED** [review]

File: src/test/ui/consts/offset_ub.rs
Comment (RalfJung) [reply src/test/ui/consts/offset_ub.rs:16]: That just checks that *any one* of these constants raises an error. We should make sure *all* of them do.

Comment (oli-obk) [reply src/test/ui/consts/offset_ub.rs:16]: I'm in full agreement. That's what I meant by `Still, annotations are better so you see something directly in the test`

Comment (josephlr) [reply src/test/ui/consts/offset_ub.rs:16]: This makes sense, I thought that for UI tests we don't need the `ERROR` annotations (because we check that the stderr matches exactly). But it looks like other tests do the same thing.

File: src/librustc_mir/interpret/intrinsics.rs
Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:438]: Fixed.

Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:438]: Fixed

**Comment by josephlr** [comment]
@RalfJung  Fixed the nits. I had to add `//~NOTE` annotations instead of `ERROR` annotations, as the const-eval errors are errors with `kind=Some(Note)` and not errors with `kind=Some(Error)`.

File: src/test/ui/consts/offset_ub.rs
Comment (RalfJung) [reply src/test/ui/consts/offset_ub.rs]: The stderr is auto-changed by `--bless`. I'd rather not rely on a reviewer realizing that a 1k line patch accidentally removed one of the errors here.

That's why we have both `ERROR` annotations and stderr files.

**Comment by RalfJung** [comment]
```
2020-05-15T20:55:09.2903132Z tidy error: /checkout/src/test/ui/consts/offset_ub.rs:12: line longer than 100 chars
2020-05-15T20:55:09.2903954Z tidy error: /checkout/src/test/ui/consts/offset_ub.rs:16: line longer than 100 chars
```
In tests I usually add the ignore-line-length thing for tidy rather than breaking the lines...^^

File: src/test/ui/consts/offset_ub.rs
Comment (RalfJung) [inline src/test/ui/consts/offset_ub.rs:7]: There are also errors on the same line... so why would `ERROR` not work?

Comment (RalfJung) [reply src/test/ui/consts/offset_ub.rs:7]: Oh, the error points to inside libcore, I see. Makes sense.

**Comment by josephlr** [comment]
@RalfJung @oli-obk I also had to change the `miri_unleashed/ptr_arith.rs` tests, as pointer arithmetic (but not pointer comparison or pointer-to-int casting) now works with `-Zunleash-the-miri-inside-of-you`. All the tests are now passing.

File: src/test/ui/consts/miri_unleashed/ptr_arith.rs
Comment (RalfJung) [inline src/test/ui/consts/miri_unleashed/ptr_arith.rs:25]: I'd rather change this to something we still do not permit... what about
```rust
let _v = (x as usize) + 0;
```

Comment (josephlr) [reply src/test/ui/consts/miri_unleashed/ptr_arith.rs:25]: I mean isn't this just exactly what we're doing above in `INT_PTR_ARITH`?

Comment (RalfJung) [reply src/test/ui/consts/miri_unleashed/ptr_arith.rs:25]: Oh. Yes. Well, one is a cast and one a transmute, might be worth having both? The `offset`-based test should just disappear I think, after all the entire point here is to implement those methods.

Comment (oli-obk) [reply src/test/ui/consts/miri_unleashed/ptr_arith.rs:25]: Alternatively you can switch to testing the `arith_offset` intrinsic, which is still not const

Comment (josephlr) [reply src/test/ui/consts/miri_unleashed/ptr_arith.rs]: Done

**Comment by oli-obk** [comment]
Oops, touchpad misfire

@bors r+

r? @oli-obk

**Comment by RalfJung** [comment]
msys2 failure
@bors retry

**Comment by RalfJung** [comment]
Failed in https://github.com/rust-lang/rust/pull/72416#issuecomment-632014546
@bors r-

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:289]:
```
error[E0599]: no function or associated item named `try_from` found for type `i64` in the current scope
   --> src/librustc_mir/interpret/intrinsics.rs:292:41
    |
292 |                 let pointee_size = i64::try_from(self.layout_of(pointee_ty)?.size.bytes()).unwrap();
    |                                         ^^^^^^^^ function or associated item not found in `i64`
    |
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope; perhaps add a `use` for it:
            `use std::convert::TryFrom;`
```

Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs:289]: Probably you need to rebase.

Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:297]: Fixed, tests should be passing now.

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:453]: Could this `unwrap` here ever actually trigger and ICE? If not, why not?

Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:453]: Good catch this can trigger an ICE. I'll add a test case and fix this.

**Comment by josephlr** [comment]
@RalfJung I fixed the potential ICE and added a test for the edge case I missed

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:454]: Please avoid `as`, it might lose information.

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:457]: Same here, please avoid `as`.

Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs:457]: In fact `offset_bytes` is negative here so what is even going on?

File: src/test/ui/consts/offset_ub.stderr
Comment (RalfJung) [inline src/test/ui/consts/offset_ub.stderr:166]: @oli-obk any idea why the span is so bad here? Shouldn't it point at just the function call?

File: src/librustc_mir/interpret/intrinsics.rs
Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:457]: This was based on how we handle computing `abs(n: isize) -> usize` with `n` negative elsewhere in `librustc_middle`: https://github.com/rust-lang/rust/blob/9eedd138ee22147111a885d6948fb050d9849bf4/src/librustc_middle/mir/interpret/pointer.rs#L52-L54

Using `as` to convert `i64` to `u64` wraps (i.e. takes the value mod 2^64) so doing a wrapping subtract after that operation computes the right value. 

It's possible to do this without an `as`, but all the alternatives are way less readable.

Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:457]: This could also be written as `offset_bytes.wrapping_abs() as u64`, which look simpler but hides the subtleties. Namely that `i64::MAX.wrapping_abs() == i64::MAX`, but then `i64::MAX as u64` then wraps to `1 << 63`, the correct value.

File: src/librustc_mir/interpret/intrinsics.rs
Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs]: Fixed to use `u64::try_from(offset_bytes).unwrap()`

File: src/test/ui/consts/offset_ub.stderr
Comment (josephlr) [reply src/test/ui/consts/offset_ub.stderr:166]: This looks to be related to the type of error that eventually bubbles up to the user. `any use of this value will cause an error` is a an error about the _value_, so that's what gets highlighted. I agree that it looks weird (though it does seem consistent with the other test cases).

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs]: Yeah I think I understood it now... could you please factor out that non-wrapping `abs` operation of type `fn(i64) -> u64` into a separate function, with suitable comments? I think that would make the code here much easier to follow.

Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs]: > Namely that i64::MAX.wrapping_abs() == i64::MAX, but then i64::MAX as u64 then wraps to 1 << 63, the correct value.

I think you mean `MIN` here?

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:456]: `MAX` is positive so it wouldn't even end up in this branch... do you mean `MIN`?

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs]: > This was based on how we handle computing abs(n: isize) -> usize with n negative elsewhere in librustc_middle: 

If we also have such code there, even more of a reason to make that `abs` function a helper function that we can share. This should probably be in libstd, but until then... maybe add it as `uabs` for the bottom of `librustc_middle/mir/interpret/mod.rs` or so?

File: src/librustc_mir/interpret/intrinsics.rs
Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs]: Yes, I meant `MIN`, this should now be correct in the `uabs` function.

File: src/librustc_mir/interpret/intrinsics.rs
Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs]: I added `uabs` to `librustc_middle/mir/interpret/mod.rs`. I wasn't able to find any other obvious call sites, though I'm sure some more exist somewhere in the tree.

**Comment by RalfJung** [comment]
Yeah this looks great. :) Thanks!
@bors r=oli-obk,RalfJung

**Comment by RalfJung** [comment]
```
66	LL |         intrinsics::offset(self, count)
67	   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
68	   |         |
-	   |         inbounds pointer arithmetic: overflow computing offset
+	   |         overflowing in-bounds pointer arithmetic
```
Looks like on 32bit the error message is different.

**Comment by josephlr** [comment]
> Looks like on 32bit the error message is different.

I guess this makes sense as all the pointer math is always done with i64 and u64. Is there a way to mark a UI test as 64 but only?

**Comment by bjorn3** [comment]
Try using `isize` and `usize`. Those are integers of the same size as a pointer.

**Comment by josephlr** [comment]
> Try using `isize` and `usize`. Those are integers of the same size as a pointer.

This won't actually work as using `isize`/`usize` are the pointer size of the host, not the target.

**Comment by RalfJung** [comment]
UB detection shouldn't depend on the target pointer size... so ideally we'd make this consistent.

The pointer offset methods in `librustc_middle/mir/interpret/pointer.rs` are all written to realize this.

**Comment by RalfJung** [comment]
In fact the two error messages are almost the same. So probably we should just adjust one of the error sites to use the same message as the other.

**Comment by RalfJung** [comment]
@josephlr you can run the 32bit test suite yourself by passing `--target i686-unknown-linux-gnu` (at least on Linux this works well if you have `gcc-multilib` installed).

**Comment by josephlr** [comment]
> @josephlr you can run the 32bit test suite yourself by passing `--target i686-unknown-linux-gnu` (at least on Linux this works well if you have `gcc-multilib` installed).

Ya I was able get the tests working (and I can see the error message issues). I think the right approach is to be more strict in the order of the UB checks.

Right now we:

1. Compute `offset_bytes`, checking for overflow of `i64`
2. Compute the `offset_pointer` via `ptr_signed_offset`, which checks that the final offset fits in a target-sized pointer.

On 32-bit, if `offset_bytes` overflows an `i32` but not an `i64`, we catch the UB in step (2) instead of step (1). I think we need to make the check in (1) more strict, making sure that `offset_bytes` doesn't overflow a _host `isize`_. Then everything should just work.

**Comment by RalfJung** [comment]
> On 32-bit, if offset_bytes overflows an i32 but not an i64, we catch the UB in step (2) instead of step (1). I think we need to make the check in (1) more strict, making sure that offset_bytes doesn't overflow a target isize. Then everything should just work.

We could do that, but it still seems odd to have almost but not quite identical error messages in both cases. IMO it would make more sense to just say "overflowing in-bounds pointer arithmetic" either way.

**Comment by josephlr** [comment]
> > On 32-bit, if offset_bytes overflows an i32 but not an i64, we catch the UB in step (2) instead of step (1). I think we need to make the check in (1) more strict, making sure that offset_bytes doesn't overflow a target isize. Then everything should just work.
> 
> We could do that, but it still seems odd to have almost but not quite identical error messages in both cases. IMO it would make more sense to just say "overflowing in-bounds pointer arithmetic" either way.

Agreed, while I implemented the checks as described above, the error is always `PointerArithOverflow`. Everything looks good now, and the tests seem to pass on 64 and 32 bit.

File: src/librustc_mir/interpret/intrinsics.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:448]: Is there any observable change in behavior from this extra check? If not, there should be at least a comment explaining why we do it.

I am not entirely sure it is worth it, given that `ptr_signed_offset` already has to check this, too.

Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:448]: So `ptr_signed_offset` does not catch the `isize` overflow if `offset_bytes` is between `isize::MAX` and `usize::MAX`. The UB is still caught in that case, but just in the memory check below

I'll add a comment explaining why this check exists.

Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs:449]: > So ptr_signed_offset does not catch the isize overflow if offset_bytes is between isize::MAX and usize::MAX.

That seems like a serious footgun, good catch!

If we decide that it is up to the caller to ensure that the input to these ptr offset methods is in the range of the target pointer type, could you add appropriate assertions? Alternatively we should make those methods properly detect overflow even for these inputs.

Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:449]: Like a `debug_assert` in libcore?

Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs:449]: In libcore? Now I am very confused.

I was more thinking of librustc_middle/mir/interpret/pointer.rs. And I'd make it proper `assert!` unless we have data that it actually slows down anything.

Comment (josephlr) [reply src/librustc_mir/interpret/intrinsics.rs:449]: I just added checks to `PointerArithmetic`'s `offset` and `signed_offset` methods to just detect overflow of the offset in these cases (and return `PointerArithOverflow`). This should make doing the right thing easier and means we don't need this check in `librustc_mir`.

File: src/librustc_middle/mir/interpret/pointer.rs
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/pointer.rs:71]: This doesn't seem right -- `overflowing_offset` will still return `false` if such an overflow occurs, right?

I think the check should probably happen in the `overflowing_*` methods instead.

Comment (josephlr) [reply src/librustc_middle/mir/interpret/pointer.rs:71]: You're right, I'll add the checks to `overflowing_offset` and `overflowing_signed_offset`.

Comment (josephlr) [reply src/librustc_middle/mir/interpret/pointer.rs]: Done

**Comment by RalfJung** [comment]
Awesome!
@bors r=oli-obk,RalfJung

**Pull request merged**
